File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    const hash = {};
    phone_book.forEach((v) => hash[v] = true);

    for (const number of phone_book) {
        for (let i = 0; i < number.length; i++) {
            const sliced = number.slice(0, i + 1)
            hash[number] = false;
            if (hash[sliced]) return false;
            hash[number] = true;
        }
    }
    return true;
}

// function solution(phone_book) {
//     return !phone_book.sort().some((_, i) => {
//         if (i === phone_book.length - 1) return false;
        
//         return phone_book[i+1].startsWith(phone_book[i]);
//     })
// }

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    # 먼저 리스트를 정렬
    phone_book.sort()
    
    # 정렬된 리스트에서 인접한 전화번호만 비교
    for i in range(len(phone_book) - 1):
        # 인접한 번호들이 접두어 관계에 있는지 확인
        if phone_book[i+1].startswith(phone_book[i]):
            return False
    return True
